lib/core/diagnostics.js

```diff
@@ -0,0 +1,209 @@
+const diagnosticsChannel = require('diagnostics_channel')
+const util = require('util')
+
+const undiciDebugLog = util.debuglog('undici')
+const fetchDebuglog = util.debuglog('fetch')
+const websocketDebuglog = util.debuglog('websocket')
+let isClientSet = false
+const channels = {
+  // Client
+  beforeConnect: diagnosticsChannel.channel('undici:client:beforeConnect'),
+  connected: diagnosticsChannel.channel('undici:client:connected'),
+  connectError: diagnosticsChannel.channel('undici:client:connectError'),
+  sendHeaders: diagnosticsChannel.channel('undici:client:sendHeaders'),
+  // Request
+  create: diagnosticsChannel.channel('undici:request:create'),
+  bodySent: diagnosticsChannel.channel('undici:request:bodySent'),
+  headers: diagnosticsChannel.channel('undici:request:headers'),
+  trailers: diagnosticsChannel.channel('undici:request:trailers'),
+  error: diagnosticsChannel.channel('undici:request:error'),
+  // WebSocket
+  open: diagnosticsChannel.channel('undici:websocket:open'),
+  close: diagnosticsChannel.channel('undici:websocket:close'),
+  socketError: diagnosticsChannel.channel('undici:websocket:socket_error'),
+  ping: diagnosticsChannel.channel('undici:websocket:ping'),
+  pong: diagnosticsChannel.channel('undici:websocket:pong')
+}
+
+if (undiciDebugLog.enabled || fetchDebuglog.enabled) {
+  const debuglog = fetchDebuglog.enabled ? fetchDebuglog : undiciDebugLog
```

Comment:
I'm having a little bit of a hard time figuring out if we should provide different debug logs; overall `websocket` is kind of straightforward, but for `fetch` not so much.

My initial idea was to have `fetch`, `undici`, and `websocket` as possible values.
If `fetch` is provided, it takes precedence, otherwise fallbacks to `undici`. `websocket` has a similar thing but is quite simpler.

Suggestions, ideas? @mcollina @Qard @KhafraDev

Reply:
Example:
```shell
NODE_DEBUG=undici,fetch,websocket node playground.js

FETCH 15809: connecting to mirror-cdn.xtom.com using https:h1
FETCH 15809: connected to mirror-cdn.xtom.com using https:h1
FETCH 15809: sending request to GET https://mirror-cdn.xtom.com//ubuntu/dists/xenial/Release.gpg
FETCH 15809: received response to GET https://mirror-cdn.xtom.com//ubuntu/dists/xenial/Release.gpg - HTTP 200
FETCH 15809: trailers received from GET https://mirror-cdn.xtom.com//ubuntu/dists/xenial/Release.gpg
```